src/driver-asus.c

```diff
@@ -0,0 +1,516 @@
+
+	// set extra settings
+	if (profile->rate_dirty) {
+		log_debug(device->ratbag, "Polling rate changed to %d Hz\n", profile->hz);
+		error = asus_set_polling_rate(device, profile->hz);
+		if (error)
+			return error;
+	}
+
+	// set DPIs
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (!resolution->dirty)
+			continue;
+
+		log_debug(
+			device->ratbag, "Resolution %d changed to %d\n",
+			resolution->index, resolution->dpi_x);
+
+		error = asus_set_dpi(device, resolution->index, resolution->dpi_x);
+		if (error)
+			return error;
+	}
+
+	// set LEDs
+	if (led_count) {
+		ratbag_profile_for_each_led(profile, led) {
+			if (!led->dirty)
+				continue;
+
+			log_debug(device->ratbag, "LED %d changed\n", led->index);
+			led_mode = 0;
+			for (i = 0; i < ARRAY_LENGTH(ASUS_LED_MODE); i++) {
+				if (ASUS_LED_MODE[i] == led->mode) {
+					led_mode = i;
+					break;
+				}
+			}
+
+			// convert brightness from 0-256 to 0-4
+			brightness = (uint8_t) round((double) led->brightness / 64.0);
+			error = asus_set_led(device, led->index, led_mode, brightness, led->color);
+			if (error)
+				return error;
+		}
+	}
+
+		// save profile
+		log_debug(device->ratbag, "Saving profile\n");
+		error = asus_save_profile(device);
+		if (error)
+			return error;
+	}
+
+	// back to initial profile
+	if (profile_count) {
+		log_debug(device->ratbag, "Switching back to initial profile %d\n", profile_id);
+		error = asus_set_profile(device, profile_id);
+		if (error)
+			return error;
+	}
+
+	return 0;
+}
+
+static int
+asus_driver_probe(struct ratbag_device *device)
+{
+	int error;
+	unsigned int profile_count, dpi_count, button_count, led_count;
+	struct asus_data *drv_data;
+	struct ratbag_profile *profile;
+	struct ratbag_button *button;
+	struct ratbag_resolution *resolution;
+	struct ratbag_led *led;
+
+	error = ratbag_open_hidraw(device);
+	if (error)
+		return error;
+
+	// create device state data
+	drv_data = zalloc(sizeof(*drv_data));
+	ratbag_set_drv_data(device, drv_data);
+	drv_data->is_ready = 1;
+
+	// get device properties from configuration file
+	profile_count = ratbag_device_data_asus_get_profile_count(device->data);
+	dpi_count = ratbag_device_data_asus_get_dpi_count(device->data);
+	button_count = ratbag_device_data_asus_get_button_count(device->data);
+	led_count = ratbag_device_data_asus_get_led_count(device->data);
+
+	// init profiles
+	ratbag_device_init_profiles(
+		device, max(profile_count, 1), dpi_count, button_count, led_count);
+
+	// setup profiles
+	ratbag_device_for_each_profile(device, profile) {
+		if (profile->index == 0) {
+			profile->is_active = true;
+		}
+
+		asus_setup_profile(device, profile);
+
+		ratbag_profile_for_each_button(profile, button)
+			asus_setup_button(device, button);
+
+		ratbag_profile_for_each_resolution(profile, resolution)
+			asus_setup_resolution(device, resolution);
+
+		if (!led_count)
+			continue;
+
+		ratbag_profile_for_each_led(profile, led)
+			asus_setup_led(device, led);
```

Comment:
same as above, this should be an empty loop if the led count is 0 so the extra check shouldn't be needed